### PR TITLE
fix: repel objective on segments causes slow optimization time

### PR DESCRIPTION
**Pull request by hsharriman**
# Description
Using `repel` to ensure that text labels are a certain distance from a segment causes optimization times of between 10-15 seconds. Based on the discussion in related issue #607, implemented 2 changes:
* Added a constraint `closestPtToSeg` that ensures a given point is a certain distance away from the **nearest point on a given segment**. As @hypotext mentions in the issue, this is faster because it does not involve sampling the distance at multiple points of the segment. After implementing this objective, the resample time drops from 10-15 seconds to 2-6 seconds.
* Per @joshpoll suggestions, added a constraint `nonDegenerateAngle` to push angles to be within a reasonable range (avoid 0 and 180-degree angles). At the moment, this is implemented with a cutoff, so any angles between 10 and 170 degrees have automatically satisfied this objective.

# Examples with steps to reproduce them
1. `yarn start`
2. Run `npx roger watch packages/synthesizer/__tests__/congruent.sub packages/synthesizer/__tests__/euclidean.sty  packages/synthesizer/__tests__/geometry.dsl` from penrose root.
# Checklist

- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] New and existing tests pass locally using `yarn test`
- [x] I ran `yarn docs` and there were no errors when generating the HTML site
- [x] My code follows the style guidelines of this project (e.g.: no ESLint warnings)

# Open questions
This was my first time writing a new constraint, and though everything appears to be working properly I'd definitely appreciate feedback on the implementation.
Questions that require more discussion or to be addressed in future development:

### CODE REVIEW

File: packages/core/src/contrib/Constraints.ts
Comment (joshpoll) [inline packages/core/src/contrib/Constraints.ts:264]: factor the cosine calculation into a new local variable and re-use it for L264 and L270

File: examples/geometry-domain/euclidean-simple.sty
Comment (joshpoll) [inline examples/geometry-domain/euclidean-simple.sty:153]: add back space on L153 and L154

File: packages/synthesizer/__tests__/euclidean.sty
Comment (joshpoll) [inline packages/synthesizer/__tests__/euclidean.sty:144]: add back space on L144 and L145

**Review by joshpoll — CHANGES_REQUESTED** [review]

File: packages/core/src/contrib/Constraints.ts
Comment (hypotext) [inline packages/core/src/contrib/Constraints.ts:235]: Looks mostly good, though I assumed you would make this a constraint, not an objective.

The thing to note is that `equalHard` is written as a constraint (as noted in the function comment). So I would either:

- Make `closestPtToSeg` a constraint and see how well that works, OR
- Use `equalHard` squared, and keep `closestPtToSeg` as an objective. That will give it nicer numerical properties as an energy (currently it is linear and looks like this: https://www.wolframalpha.com/input/?i=f%28x%2C+y%29+%3D+max%28x%2C+y%29+-+min%28x%2C+y%29 )

Let me know how well either option works. 

Also, I'd give your function a more descriptive name, like `pointAtDistFromClosestPt`.

**Review by hypotext — CHANGES_REQUESTED** [review]

File: packages/core/src/contrib/Constraints.ts
Comment (hsharriman) [reply packages/core/src/contrib/Constraints.ts:235]: I tried both of your suggestions, and it seems like keeping it as an objective is a bit faster and more reliable than making it a constraint. When I made it a constraint, every now and again it would take ~3minutes to optimize and end in a degenerated state such as this:
Screenshot from 2021-06-28 11-38-48
Keeping it as an objective still results in the 2-5 second time between resamples that I described in the original PR comment (as a constraint it is closer to 5 seconds on average)

**Review by hypotext — APPROVED** [review]
These changes look fine to me. @joshpoll @wodeni What do you think, is this ready to merge?

File: packages/core/src/contrib/Constraints.ts
Comment (wodeni) [inline packages/core/src/contrib/Constraints.ts:252]: It would be good if we can take the min and max angles as arguments, perhaps with defaults `10` and `170`?

**Review by wodeni — APPROVED** [review]
Looking at the code, the changes look good to me. Seems like the perf problem is still there (?), but I think it's good to merge as long as @hypotext and @joshpoll think these functions are better alternatives to `repel`

File: packages/core/src/contrib/Constraints.ts
Comment (joshpoll) [inline packages/core/src/contrib/Constraints.ts:265]: maybe `pointLineDist` or something similar would be a better name? point-line distance is canonically the shortest distance from the point to the line

**Review by joshpoll — CHANGES_REQUESTED** [review]

File: packages/core/src/contrib/Constraints.ts
Comment (hsharriman) [reply packages/core/src/contrib/Constraints.ts]: I added a `range` variable that defines how many degrees from 0 or 180 are considered unsafe (it's 10 by default, so angles <10 degrees or >170 degrees are in the "unsafe" range)

File: packages/core/src/contrib/Constraints.ts
Comment (joshpoll) [inline packages/core/src/contrib/Constraints.ts:267]: Apologies for the back and forth here, but if you could switch the arguments to this objective so the first argument is the point and the second argument is the line that would be awesome. Also please test if the line `isLineLike` and throw an error if it's not. (check the other constraints and objectives to see how this is done)

**Review by joshpoll — CHANGES_REQUESTED** [review]
overall lgtm once this change is incorporated I will approve!

**Review by joshpoll — APPROVED** [review]

**Comment by hypotext** [comment]
Forgot @joshpoll didn't have merge permission. Done now, thanks @hsharriman!

**Pull request merged**
